{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Jul 21 18:55:02 PDT 2021\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "# Notebook 1: Downloading sequencing data. \n",
    "#### This notebook was written by Victoria Jorgensen. \n",
    "Questions can be sent to jorgensen at caltech.edu. \n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import all packages (some not necessary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "\n",
    "%config InlineBackend.figure_format = 'retina'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading in supplemental data: \n",
    "You will first need to download all supplementary data (meta data and gene lists) from GitHub repository [hEP-structures_MZG](https://github.com/vjorgensen/hEP-structures_MZG).  Sequencing data above was added to the \"data\" folder in the above github repository in a subsequent folder titled \"matrix_data\" and then further into folders \"MZG_hEP\" and \"MZG_nat\" depending on the source of each sample. This step is necessary because the sequencing files were too large to upload to GitHub. \n",
    "\n",
    "## Loading in sequencing data:  \n",
    "This notebook will help you read in all sequencing data generated in this study for 2D hEPSCs, Day 5 hEP-structures, Day 6 hEP-structures, and Day 5/6 natural embryos. To use this notebook, you will need to download the following from GEO at accession number [GSE178326](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178326). Download \"GSE178326_RAW.tar\" in the supplementary files for the following:\n",
    "\n",
    "For hEP-structures (put data in folder \"data/matrix_data/MZG_hEP\"):\n",
    "- GSM5387817_EPSC_matrix.mtx, \n",
    "- GSM5387817_EPSC_barcodes.tsv\n",
    "- GSM5387817_EPSC_features.tsv\n",
    "- GSM5387817_EPSC_LMO-tags.csv\n",
    "\n",
    "For natural embryos (put data in folder \"data/matrix_data/MZG_nat\"): \n",
    "- GSM5387818_natural_matrix.mtx, \n",
    "- GSM5387818_natural_barcodes.tsv\n",
    "- GSM5387818_natural_features.tsv\n",
    "\n",
    "\n",
    "\n",
    "**Note: You will have to rename the above files to \"matrix.mtx.gz\", \"barcodes.tsv.gz\", and \"features.tsv.gz\" in \"MZG_hEP\" and \"MZG_nat\" to use the following code!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data from hEP-structures\n",
    "Goetz_D = sc.read_10x_mtx(\"../data/matrix_data/MZG_hEP/\")\n",
    "\n",
    "# Data from D5/D6 natural embryos\n",
    "Goetz_D2 = sc.read_10x_mtx(\"../data/matrix_data/MZG_nat/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Concatenate the two data frames. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = Goetz_D.concatenate(Goetz_D2, index_unique=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Organize the data by adding in gene_name column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.3</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>RP11-34P13.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC233755.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC233755.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC240274.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC213203.1</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC213203.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM231B</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>FAM231B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33694 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_ids    feature_types     gene_name\n",
       "RP11-34P13.3  ENSG00000243485  Gene Expression  RP11-34P13.3\n",
       "FAM138A       ENSG00000237613  Gene Expression       FAM138A\n",
       "OR4F5         ENSG00000186092  Gene Expression         OR4F5\n",
       "RP11-34P13.7  ENSG00000238009  Gene Expression  RP11-34P13.7\n",
       "RP11-34P13.8  ENSG00000239945  Gene Expression  RP11-34P13.8\n",
       "...                       ...              ...           ...\n",
       "AC233755.2    ENSG00000277856  Gene Expression    AC233755.2\n",
       "AC233755.1    ENSG00000275063  Gene Expression    AC233755.1\n",
       "AC240274.1    ENSG00000271254  Gene Expression    AC240274.1\n",
       "AC213203.1    ENSG00000277475  Gene Expression    AC213203.1\n",
       "FAM231B       ENSG00000268674  Gene Expression       FAM231B\n",
       "\n",
       "[33694 rows x 3 columns]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make gene name column\n",
    "adata.var[\"gene_name\"] = adata.var.index\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load in labels for samples based on barcodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGATGAC11</th>\n",
       "      <td>unknown</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTAAAGG11</th>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT11</th>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA11</th>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACCGTGC11</th>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCACAGTCCGGACT-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCATGCATCCTCAC-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCAGGAGCAAA-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGACGGTTG-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCATTTCA-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6231 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sample_id  sample_number\n",
       "AAACCCAAGCGATGAC11   unknown              0\n",
       "AAACCCAAGGTAAAGG11     D6-R3              6\n",
       "AAACCCAAGTTGAAGT11     D6-R1              4\n",
       "AAACCCACAAATCCCA11     D6-R3              6\n",
       "AAACCCACAACCGTGC11     D6-R1              4\n",
       "...                      ...            ...\n",
       "TTTCACAGTCCGGACT-1       Nat              7\n",
       "TTTCATGCATCCTCAC-1       Nat              7\n",
       "TTTGATCAGGAGCAAA-1       Nat              7\n",
       "TTTGGTTAGACGGTTG-1       Nat              7\n",
       "TTTGGTTTCCATTTCA-1       Nat              7\n",
       "\n",
       "[6231 rows x 2 columns]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Barcodes/LMOs for EPSCs, D5 hEP-structures, and D6 hEP-structures.  (Multiplexed samples, hence LMO tags)\n",
    "tags_ep = pd.read_csv(\"../data/matrix_data/MZG_hEP/GSM5387817_EPSC_LMO-tags.csv\")\n",
    "\n",
    "# Barcodes/LMOs for EPSCs, D5 hEP-structures, and D6 hEP-structures.  (Multiplexed samples, hence LMO tags)\n",
    "tags_nat = pd.read_csv(\"../data/matrix_data/MZG_nat/barcodes.tsv.gz\", names=[\"cell_barcode\", \"sample_id\", \"sample_number\"], header=None)\n",
    "tags_nat[\"sample_id\"] = \"Nat\"\n",
    "tags_nat[\"sample_number\"] = 7\n",
    "\n",
    "# Combine the two dataframes, note that order must match order of adata file. (EPSC then natural samples)\n",
    "tag = (tags_ep, tags_nat)\n",
    "tags = pd.concat(tag, ignore_index=True)\n",
    "tags = tags.set_index(\"cell_barcode\").rename_axis(index=None, axis=1)\n",
    "tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_number</th>\n",
       "      <th>cell_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGATGAC11</th>\n",
       "      <td>unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTAAAGG11</th>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT11</th>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA11</th>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACCGTGC11</th>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCACAGTCCGGACT-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCATGCATCCTCAC-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCAGGAGCAAA-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGACGGTTG-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCATTTCA-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6231 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sample_id  sample_number cell_group\n",
       "AAACCCAAGCGATGAC11   unknown              0    unknown\n",
       "AAACCCAAGGTAAAGG11     D6-R3              6         D6\n",
       "AAACCCAAGTTGAAGT11     D6-R1              4         D6\n",
       "AAACCCACAAATCCCA11     D6-R3              6         D6\n",
       "AAACCCACAACCGTGC11     D6-R1              4         D6\n",
       "...                      ...            ...        ...\n",
       "TTTCACAGTCCGGACT-1       Nat              7        Nat\n",
       "TTTCATGCATCCTCAC-1       Nat              7        Nat\n",
       "TTTGATCAGGAGCAAA-1       Nat              7        Nat\n",
       "TTTGGTTAGACGGTTG-1       Nat              7        Nat\n",
       "TTTGGTTTCCATTTCA-1       Nat              7        Nat\n",
       "\n",
       "[6231 rows x 3 columns]"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate 'cell_group' column to distinguish natural embryo, 2D EPSCs, D5 hEP, and D6 hEP.\n",
    "# Note that 'sample_id' distinguishes between replicates.\n",
    "tags['cell_group'] = [tags['sample_id'][i].split('-')[0] for i in range(len(tags))]\n",
    "tags.index.name = None\n",
    "\n",
    "tags"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add this sample data to adata. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_number</th>\n",
       "      <th>cell_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGATGAC11</th>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTAAAGG11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACCGTGC11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCACAGTCCGGACT-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCATGCATCCTCAC-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCAGGAGCAAA-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGACGGTTG-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCATTTCA-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6231 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   batch sample_id  sample_number cell_group\n",
       "AAACCCAAGCGATGAC11     0   unknown              0    unknown\n",
       "AAACCCAAGGTAAAGG11     0     D6-R3              6         D6\n",
       "AAACCCAAGTTGAAGT11     0     D6-R1              4         D6\n",
       "AAACCCACAAATCCCA11     0     D6-R3              6         D6\n",
       "AAACCCACAACCGTGC11     0     D6-R1              4         D6\n",
       "...                  ...       ...            ...        ...\n",
       "TTTCACAGTCCGGACT-1     1       Nat              7        Nat\n",
       "TTTCATGCATCCTCAC-1     1       Nat              7        Nat\n",
       "TTTGATCAGGAGCAAA-1     1       Nat              7        Nat\n",
       "TTTGGTTAGACGGTTG-1     1       Nat              7        Nat\n",
       "TTTGGTTTCCATTTCA-1     1       Nat              7        Nat\n",
       "\n",
       "[6231 rows x 4 columns]"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add meta data to adata.obs\n",
    "adata.obs = pd.concat([adata.obs, tags], axis=1)\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove all \"unknown\" cells\n",
    "Some samples were labeled \"unknown\" due to duplicates from multiplexing/ambiguous LMO tagging/etc. The following code removes those cells from further analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_number</th>\n",
       "      <th>cell_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTAAAGG11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R3</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACCGTGC11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R1</td>\n",
       "      <td>4</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACAACATC11</th>\n",
       "      <td>0</td>\n",
       "      <td>D6-R2</td>\n",
       "      <td>5</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCACAGTCCGGACT-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCATGCATCCTCAC-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCAGGAGCAAA-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGACGGTTG-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCATTTCA-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Nat</td>\n",
       "      <td>7</td>\n",
       "      <td>Nat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4840 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   batch sample_id  sample_number cell_group\n",
       "AAACCCAAGGTAAAGG11     0     D6-R3              6         D6\n",
       "AAACCCAAGTTGAAGT11     0     D6-R1              4         D6\n",
       "AAACCCACAAATCCCA11     0     D6-R3              6         D6\n",
       "AAACCCACAACCGTGC11     0     D6-R1              4         D6\n",
       "AAACCCACACAACATC11     0     D6-R2              5         D6\n",
       "...                  ...       ...            ...        ...\n",
       "TTTCACAGTCCGGACT-1     1       Nat              7        Nat\n",
       "TTTCATGCATCCTCAC-1     1       Nat              7        Nat\n",
       "TTTGATCAGGAGCAAA-1     1       Nat              7        Nat\n",
       "TTTGGTTAGACGGTTG-1     1       Nat              7        Nat\n",
       "TTTGGTTTCCATTTCA-1     1       Nat              7        Nat\n",
       "\n",
       "[4840 rows x 4 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Removes cells with unknown sample-id (i.e. cells with ambiguous barcoding)\n",
    "adata = adata[adata.obs['sample_id']!= \"unknown\"] \n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make sample titles more clear."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "# Give labels a more descriptive name \n",
    "adata.obs[\"sample_group\"] = \" \"\n",
    "adata.obs.loc[(adata.obs[\"cell_group\"]==\"Nat\"), \"sample_group\"] = \"Natural human embryo\"\n",
    "adata.obs.loc[(adata.obs[\"cell_group\"]==\"D5\"), \"sample_group\"] = \"Day 5 hEP-structures\"\n",
    "adata.obs.loc[(adata.obs[\"cell_group\"]==\"D6\"), \"sample_group\"] = \"Day 6 hEP-structures\"\n",
    "adata.obs.loc[(adata.obs[\"cell_group\"]==\"2D\"), \"sample_group\"] = \"hEPSCs in 2D\"\n",
    "# adata.obs.loc[(adata.obs[\"cell_group\"]==\"unknown\"), \"sample_group\"] = \"unknown\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove superfluous columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>cell_group</th>\n",
       "      <th>sample_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTAAAGG11</th>\n",
       "      <td>D6-R3</td>\n",
       "      <td>D6</td>\n",
       "      <td>Day 6 hEP-structures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT11</th>\n",
       "      <td>D6-R1</td>\n",
       "      <td>D6</td>\n",
       "      <td>Day 6 hEP-structures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA11</th>\n",
       "      <td>D6-R3</td>\n",
       "      <td>D6</td>\n",
       "      <td>Day 6 hEP-structures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACCGTGC11</th>\n",
       "      <td>D6-R1</td>\n",
       "      <td>D6</td>\n",
       "      <td>Day 6 hEP-structures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACAACATC11</th>\n",
       "      <td>D6-R2</td>\n",
       "      <td>D6</td>\n",
       "      <td>Day 6 hEP-structures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCACAGTCCGGACT-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>Nat</td>\n",
       "      <td>Natural human embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCATGCATCCTCAC-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>Nat</td>\n",
       "      <td>Natural human embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCAGGAGCAAA-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>Nat</td>\n",
       "      <td>Natural human embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGACGGTTG-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>Nat</td>\n",
       "      <td>Natural human embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCATTTCA-1</th>\n",
       "      <td>Nat</td>\n",
       "      <td>Nat</td>\n",
       "      <td>Natural human embryo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4840 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sample_id cell_group          sample_group\n",
       "AAACCCAAGGTAAAGG11     D6-R3         D6  Day 6 hEP-structures\n",
       "AAACCCAAGTTGAAGT11     D6-R1         D6  Day 6 hEP-structures\n",
       "AAACCCACAAATCCCA11     D6-R3         D6  Day 6 hEP-structures\n",
       "AAACCCACAACCGTGC11     D6-R1         D6  Day 6 hEP-structures\n",
       "AAACCCACACAACATC11     D6-R2         D6  Day 6 hEP-structures\n",
       "...                      ...        ...                   ...\n",
       "TTTCACAGTCCGGACT-1       Nat        Nat  Natural human embryo\n",
       "TTTCATGCATCCTCAC-1       Nat        Nat  Natural human embryo\n",
       "TTTGATCAGGAGCAAA-1       Nat        Nat  Natural human embryo\n",
       "TTTGGTTAGACGGTTG-1       Nat        Nat  Natural human embryo\n",
       "TTTGGTTTCCATTTCA-1       Nat        Nat  Natural human embryo\n",
       "\n",
       "[4840 rows x 3 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs = adata.obs.drop(columns=[\"batch\", \"sample_number\"])\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save adata file to be used for nb2_data-analaysis.ipynb in matrix_data folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample_id' as categorical\n",
      "... storing 'cell_group' as categorical\n",
      "... storing 'sample_group' as categorical\n",
      "... storing 'feature_types' as categorical\n"
     ]
    }
   ],
   "source": [
    "# Save adata.\n",
    "adata.write(\"../data/matrix_data/adata.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The watermark extension is already loaded. To reload it, use:\n",
      "  %reload_ext watermark\n",
      "CPython 3.7.10\n",
      "IPython 7.22.0\n",
      "\n",
      "numpy 1.19.4\n",
      "pandas 1.1.4\n",
      "scanpy 1.7.2\n",
      "anndata 0.7.6\n",
      "jupyterlab 3.0.11\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -p numpy,pandas,scanpy,anndata,jupyterlab"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
